sandbox/windowed_spatial_avg.jl

```diff
@@ -0,0 +1,101 @@
+using Printf
+using Revise
+using Oceananigans
+using Oceananigans.Fields
+using Oceananigans.OutputWriters
+using Oceananigans.Advection
+using Oceananigans.Utils
+
+#++++ Model set-up
+Nx = Ny = Nz = 8
+Lx = Ly = Lz = 1
+N² = 1e-4 # s⁻²
+
+grid = RegularRectilinearGrid(size=(Nx, Ny, Nz), extent=(Lx, Ly, Lz), topology=(Periodic, Periodic, Periodic))
+
+model = IncompressibleModel(
+                   grid = grid,
+              advection = UpwindBiasedFifthOrder(),
+            timestepper = :RungeKutta3,
+                closure = IsotropicDiffusivity(ν=1e-4, κ=1e-4),
+               coriolis = FPlane(f=1e-4),
+                tracers = (:b,), # P for Plankton
+               buoyancy = BuoyancyTracer(),
+)
+println()
+println(model)
+println()
+#-----
+
+#++++ Initial conditions
+u0(x, y, z) = y<Ly/2 ? 0.5 : 0.7
+v0(x, y, z) = 0.2
+w0(x, y, z) = 0.2*y
+
+set!(model, u=u0, v=v0, w=w0)
+#-----
+
+
+#++++
+struct WindowedSpatialAverage{F, S, D}
+          field :: F
+   field_slicer :: S
+           dims :: D
+end
+WindowedSpatialAverage(field; dims, field_slicer=FieldSlicer()) = WindowedSpatialAverage(field, field_slicer, dims)
+
+using Oceananigans.OutputWriters: slice_parent
+using Statistics: mean
+function (wsa::WindowedSpatialAverage)(model)
+    compute!(wsa.field)
+    window = slice_parent(wsa.field_slicer, wsa.field)
+    return dropdims(mean(window, dims=wsa.dims), dims=wsa.dims)
+end
+
+using NCDatasets: defVar
+using Oceananigans.Fields: reduced_location
+import Oceananigans.OutputWriters: xdim, ydim, zdim, define_output_variable!
+
+function define_output_variable!(dataset, 
+                                 wtsa::Union{WindowedSpatialAverage, WindowedTimeAverage{<:WindowedSpatialAverage}}, 
+                                 name, array_type, compression, attributes, dimensions)
+    if wtsa isa WindowedSpatialAverage
+        wsa = wtsa
+    elseif wtsa isa WindowedTimeAverage
+        wsa = wtsa.operand
+    else
+        throw("Wrong type for windowed averaged")
+    end
```

Comment:
This will work. Another solution using dispatch (which has the advantage of being more extensible is to use something like)

```julia
unwrap_time_average(wta::WindowedTimeAverage) = wta.operand
unwrap_time_average(wsa::WindowedSpatialAverage) = wsa
```

and then in the body of `define_output_variable!`:

```julia
wsa = unwrap_time_average(wtsa)
```

You can also define a fallback that throws an error.

Purely for the sake of example, another solution would be to define appropriate `location` and `dims` functions (since I think this is what interferes with generalizability):

```julia
location(wsa::WindowedSpatialAverage) = location(wsa.field)
location(wta::WindowedTimeAverage{<:WindowedSpatialAverage}) = location(wta.operand)

dims(wsa::WindowedSpatialAverage) = wsa.dims
dims(wta::WindowedTimeAverage{<:WindowedSpatialAverage}) = dims(wta.operand)
```

Then below use

```julia
LX, LY, LZ = reduced_location(location(wsa), dims=dims(wsa))
```

Reply:
Note that for the above solution to work `location` must be imported:

```julia
import Oceananigans.Fields: location
```

I don't think we have a `dims` function yet (I guess maybe `reduced_dims` could be a better / more verbose name).